cmd/prometheus/main.go

```diff
@@ -288,6 +292,17 @@ func main() {
 
 	var g group.Group
 	{
+		g.Add(
+			func() error {
+				return discoveryManager.Run()
+			},
+			func(err error) {
+				level.Info(logger).Log("msg", "Stopping target manager...")
+				cancelCtx()
```

Comment:
A single actor canceling a context used across components doesn't seem right. Each component should probably have it's own context with a dedicated cancel function.

Reply:
ok , will change.
would be interested if you can think of  an example when using a single context can cause an issue.

Reply:
Cannot think of something for this particular case. But as soon as you inspect the interrupt error and decide to handle it specifically, any dependency of one actor on the interrupt function on another will get you in trouble I'd imagine.

Seems a good idea in general to avoid it all together.